File: src/main/java/io/scriptor/http/HTTPServer.java
package io.scriptor.http;

import io.scriptor.annotation.*;
import io.scriptor.http.result.HTTPResult;
import io.scriptor.log.Log;
import io.scriptor.type.IConverter;
import io.scriptor.type.TypeRef;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.net.ServerSocketFactory;
import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocket;
import java.io.ByteArrayInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.net.ServerSocket;
import java.net.Socket;
import java.security.*;
import java.security.cert.CertificateException;
import java.util.*;
import java.util.concurrent.*;

public class HTTPServer implements AutoCloseable {

    private record RouteBundle(
            @NotNull Object instance,
            @NotNull Method callee,
            @NotNull HTTPRoute route,
            @NotNull String accept,
            @NotNull String result
    ) {
    }

    private final boolean tls;

    private final ServerSocket serverSocket;

    private final Map<HTTPMethod, List<RouteBundle>> routes = new HashMap<>();
    private final Map<Type, Map<Type, IConverter<?, ?>>> converters = new HashMap<>();

    private final BlockingQueue<Runnable> workQueue = new ArrayBlockingQueue<>(256);
    private final Executor executor = new ThreadPoolExecutor(10,
                                                             10,
                                                             10,
                                                             TimeUnit.MINUTES,
                                                             workQueue);

    public HTTPServer(
            final int port,
            final boolean enableTLS,
            final @Nullable String keystoreFilename,
            final @Nullable String keystorePassphrase
    )
            throws CertificateException,
                   IOException,
                   KeyManagementException,
                   KeyStoreException,
                   NoSuchAlgorithmException,
                   UnrecoverableKeyException {

        tls = enableTLS;

        final ServerSocketFactory serverSocketFactory;
        if (enableTLS) {
            final var passphrase = Objects.requireNonNull(keystorePassphrase).toCharArray();
            final var keystore   = KeyStore.getInstance("JKS");
            keystore.load(new FileInputStream(Objects.requireNonNull(keystoreFilename)), passphrase);

            final var keyManagerFactory = KeyManagerFactory.getInstance("SunX509");
            keyManagerFactory.init(keystore, passphrase);

            final var sslContext = SSLContext.getInstance("TLS");
            sslContext.init(keyManagerFactory.getKeyManagers(), null, null);

            serverSocketFactory = sslContext.getServerSocketFactory();
        } else {
            serverSocketFactory = ServerSocketFactory.getDefault();
        }

        serverSocket = serverSocketFactory.createServerSocket(port);
        Log.info("HTTP%s server listening on port %d", enableTLS ? "S" : "", port);
    }

    public void registerRoute(
            final @NotNull Object instance,
            final @NotNull Method callee,
            final @NotNull Endpoint endpoint,
            final @NotNull Resource resource
    ) {
        routes.computeIfAbsent(resource.method(), _ -> new ArrayList<>())
              .add(new RouteBundle(instance,
                                   callee,
                                   new HTTPRoute(endpoint.value(), resource.path()),
                                   resource.accept(),
                                   resource.result()));
    }

    public <S, D> void registerConverter(
            final @NotNull Type source,
            final @NotNull Type destination,
            final @NotNull IConverter<S, D> converter
    ) {
        converters.computeIfAbsent(source, _ -> new HashMap<>()).put(destination, converter);
    }

    public void handleRequest() throws IOException {
        final var socket = serverSocket.accept();
        executor.execute(() -> {
            try {
                if (tls) {
                    ((SSLSocket) socket).startHandshake();
                }
                handleRequest(socket);
            } catch (final IOException e) {
                Log.trace(e);
            }
        });
    }

    public void start() throws IOException {
        while (!Thread.interrupted()) {
            handleRequest();
        }
    }

    @SuppressWarnings("unchecked")
    private <S, D> D convert(final @Nullable S object, final @NotNull Type source, final @NotNull Type destination) {
        if (IConverter.isAssignable(destination, source)) {
            return (D) object;
        }

        if (converters.containsKey(source) && converters.get(source).containsKey(destination)) {
            return ((IConverter<S, D>) converters.get(source).get(destination)).from(object);
        }

        throw new IllegalStateException("unsupported conversion from '%s' to '%s'".formatted(source, destination));
    }

    private void handleRequest(final @NotNull Socket socket) throws IOException {
        try (
                final var inputStream = socket.getInputStream();
                final var outputStream = socket.getOutputStream()
        ) {
            final var request = HTTPRequestMessage.read(inputStream);

            if (request == null) {
                Log.warning("invalid request: request is null");
                return;
            }

            Log.info("%s %s %s", request.method(), request.path(), request.protocol());

            for (final var bundle : routes.get(request.method())) {
                if (!bundle.route().matches(request.path())) {
                    continue;
                }

                final var parameterCount = bundle.callee().getParameterCount();
                final var parameters     = bundle.callee().getParameters();

                final var args = new Object[parameterCount];

                final HTTPResult<?> result;
                try {
                    for (int i = 0; i < parameterCount; ++i) {
                        final var parameter = parameters[i];

                        final Object value;
                        if (parameter.isAnnotationPresent(Body.class)) {
                            value = request.body();
                        } else if (parameter.isAnnotationPresent(Header.class)) {
                            final var name = Objects.requireNonNull(parameter.getAnnotation(Header.class)).value();
                            value = request.headers().get(name.toLowerCase());
                        } else if (parameter.isAnnotationPresent(Path.class)) {
                            final var name = Objects.requireNonNull(parameter.getAnnotation(Path.class)).value();
                            value = bundle.route().get(request.path(), name);
                        } else if (parameter.isAnnotationPresent(Query.class)) {
                            final var name   = Objects.requireNonNull(parameter.getAnnotation(Query.class)).value();
                            final var values = request.query().computeIfAbsent(name, _ -> new ArrayList<>());
                            value = parameter.getType().isArray()
                                    ? values.toArray()
                                    : !values.isEmpty()
                                      ? values.getFirst()
                                      : null;
                        } else {
                            value = null;
                        }

                        if (value != null) {
                            args[i] = convert(value, value.getClass(), parameter.getType());
                        }
                    }

                    final var object = bundle.callee().invoke(bundle.instance(), args);
                    final var type   = bundle.callee().getGenericReturnType();
                    result = convert(object, type, new TypeRef<HTTPResult<?>>() {
                    }.getType());
                } catch (final Exception e) {
                    Log.trace(e);
                    new HTTPResponseMessage("HTTP/1.1",
                                            500,
                                            "Internal Server Error - %s".formatted(e.getMessage()),
                                            new HashMap<>(),
                                            null,
                                            false).write(outputStream);
                    return;
                }

                final Map<String, String> headers = new HashMap<>(result.getHeaders());
                headers.put("Content-Type", bundle.result());
                headers.put("Connection", "Close");

                final var chunked = result.getSize() < 0;
                if (chunked) {
                    headers.put("Transfer-Encoding", "chunked");
                } else {
                    headers.put("Content-Length", Integer.toString(result.getSize()));
                }

                new HTTPResponseMessage("HTTP/1.1",
                                        result.getStatusCode(),
                                        result.getStatusText(),
                                        headers,
                                        result.getStream(),
                                        chunked).write(outputStream);
                return;
            }

            notFound().write(outputStream);
        }
    }

    private static @NotNull HTTPResponseMessage notFound() {
        final var bytes = "resource not found".getBytes();

        final Map<String, String> headers = new HashMap<>();
        headers.put("Content-Type", "text/plain");
        headers.put("Content-Length", Integer.toString(bytes.length));
        headers.put("Connection", "Close");

        final InputStream body = new ByteArrayInputStream(bytes);

        return new HTTPResponseMessage("HTTP/1.1",
                                       404,
                                       "Not Found",
                                       headers,
                                       body,
                                       false);
    }

    @Override
    public void close() throws IOException {
        serverSocket.close();
    }
}

File: server/src/main/kotlin/dev/scriptor/server/http/HTTPServer.kt
package dev.scriptor.server.http

import dev.scriptor.server.*
import dev.scriptor.server.address.AddressType.*
import dev.scriptor.server.address.normalizeIpv4
import dev.scriptor.server.address.normalizeIpv6
import dev.scriptor.server.address.normalizeName
import dev.scriptor.server.address.parseAddressType
import dev.scriptor.server.annotation.*
import dev.scriptor.server.converter.ConversionPath
import dev.scriptor.server.converter.ConversionStep
import dev.scriptor.server.converter.Converter
import dev.scriptor.server.http.result.HTTPResult
import dev.scriptor.server.http.result.HTTPResultUnit
import dev.scriptor.server.type.isAssignable
import java.io.IOException
import java.lang.AutoCloseable
import java.lang.reflect.InvocationTargetException
import java.net.InetSocketAddress
import java.nio.channels.ServerSocketChannel
import java.nio.channels.SocketChannel
import java.util.*
import java.util.concurrent.*
import java.util.logging.Logger
import kotlin.concurrent.timerTask
import kotlin.reflect.*
import kotlin.reflect.full.findAnnotation
import kotlin.reflect.full.hasAnnotation
import kotlin.reflect.full.memberProperties
import kotlin.reflect.full.starProjectedType

class HTTPServer(
    val log: Logger,
    val hostname: String,
    val port: Int,
) : AutoCloseable {

    data class Route(
        val instance: Any,
        val callee: KCallable<*>,
        val route: HTTPRoute,
        val method: HTTPMethod,
        val accept: String,
        val result: String
    ) : Comparable<Route> {

        override fun compareTo(other: Route): Int {
            return this.route.compareTo(other.route)
        }

        override fun toString(): String {
            return "$method $route : $accept -> $result"
        }
    }

    private data class Task(
        val interval: Long,
        val callee: Runnable,
        val task: TimerTask,
    )

    private val server = ServerSocketChannel.open()

    private val routes = mutableMapOf<HTTPMethod, MutableList<Route>>()
    private val converters = mutableListOf<ConversionStep>()
    private val instances = mutableListOf<Any>()
    private val injected = mutableMapOf<String, Any?>()

    private val timer = Timer()
    private val tasks = mutableMapOf<String, Task>()

    private val queue: BlockingQueue<Runnable> = ArrayBlockingQueue(256)
    private val executor: Executor = ThreadPoolExecutor(
        10,
        10,
        100,
        TimeUnit.MILLISECONDS,
        queue
    )

    private val conversionCache = mutableMapOf<Pair<KType, KType>, ConversionPath<Any, Any>>()

    private var running: Boolean = false

    init {
        server.bind(InetSocketAddress(hostname, port))

        val addressType = parseAddressType(hostname)
        val normalized = when (addressType) {
            INVALID -> throw Error("invalid hostname '$hostname'")
            IPV4 -> normalizeIpv4(hostname)
            IPV6 -> normalizeIpv6(hostname)
            NAME -> normalizeName(hostname)
        }

        log.info("server listening on ${if (':' in normalized) "[$normalized]" else normalized}:$port")
    }

    override fun close() {
        server.close()
        timer.cancel()
    }

    fun registerRoute(
        instance: Any,
        callee: KCallable<*>,
        endpoint: Endpoint,
        resource: Resource,
    ): Route {
        checkConvertible(callee.returnType, typeOf<HTTPResult<*>>())

        val route = Route(
            instance,
            callee,
            HTTPRoute(endpoint.value, resource.path),
            resource.method,
            resource.accept,
            resource.result
        )

        routes.computeIfAbsent(resource.method) { mutableListOf() } += route

        return route
    }

    fun registerConverter(
        src: KType,
        dst: KType,
        converter: Converter<*, *>,
    ) {
        converters += ConversionStep(
            src,
            dst,
            converter as Converter<Any, Any>,
        )

        injectInstance(converter)
    }

    fun registerContext(name: String, instance: Any) {
        instances += instance

        injectInstance(instance)

        inject(name, instance)
    }

    fun registerEndpoint(instance: Any) {
        instances += instance

        injectInstance(instance)
    }

    fun inject(name: String, value: Any?) {
        injected[name] = value

        for (instance in instances) {
            for (property in instance::class.memberProperties) {
                if (property !is KMutableProperty<*>) continue

                val inject = property.findAnnotation<Inject>() ?: continue
                if (inject.value != name) continue

                property.setter.call(instance, value)
            }
        }
    }

    fun registerTask(name: String, interval: Long, callee: Runnable) {
        val task = timerTask { callee.run() }
        tasks[name] = Task(interval, callee, task)

        timer.scheduleAtFixedRate(task, interval, interval)
    }

    fun cancelTask(name: String) {
        val value = tasks.remove(name) ?: return
        value.task.cancel()
    }

    fun spin() {
        val socket = server.accept()

        executor.execute {
            try {
                handle(socket)
            } catch (e: IOException) {
                log.trace(e)
            } finally {
                socket.close()
            }
        }
    }

    fun start() {
        running = true

        while (running && !Thread.interrupted()) {
            spin()
        }
    }

    fun stop() {
        running = false
    }

    private fun injectInstance(instance: Any) {
        for (property in instance::class.memberProperties) {
            if (property !is KMutableProperty<*>) continue

            if (property.hasAnnotation<Conversion>()) {
                if (property.returnType.classifier != ConversionPath::class) continue

                val src = property.returnType.arguments[0].type!!
                val dst = property.returnType.arguments[1].type!!

                property.setter.call(instance, findConversionPath(src, dst) ?: continue)
                continue
            }

            val inject = property.findAnnotation<Inject>() ?: continue
            if (inject.value !in injected) continue

            property.setter.call(instance, injected[inject.value])
        }
    }

    private fun findConversionPath(src: KType, dst: KType): ConversionPath<Any, Any>? {

        val key = Pair(src, dst)

        if (key in conversionCache) {
            return conversionCache[key]
        }

        data class Node(
            val type: KType,
            val path: List<ConversionStep>,
        )

        val queue = ArrayDeque<Node>()
        val visited = HashSet<KType>()

        queue.add(Node(src, emptyList()))

        while (queue.isNotEmpty()) {

            val current = queue.removeFirst()

            if (!visited.add(current.type)) continue

            if (isAssignable(dst, current.type)) {
                val path = ConversionPath<Any, Any>(current.path)
                conversionCache[key] = path
                return path
            }

            val edges = converters.filter { isAssignable(it.src, current.type) }

            for ((_, next, converter) in edges) {
                queue += Node(
                    next,
                    current.path + ConversionStep(
                        current.type,
                        next,
                        converter,
                    )
                )
            }
        }

        return null
    }

    private fun convertible(src: KType, dst: KType): Boolean {
        return findConversionPath(src, dst) != null
    }

    private fun checkConvertible(src: KType, dst: KType) {
        if (convertible(src, dst)) return

        throw Exception("unsupported conversion from '$src' to '$dst'")
    }

    private fun convert(value: Any, src: KType, dst: KType): Any {

        val path = findConversionPath(src, dst)
            ?: throw Exception("unsupported conversion from '$src' to '$dst'")

        return path.convert(value)
    }

    private fun handle(channel: SocketChannel) {
        val reader = HTTPRequestMessageReader(BufferedReadableByteChannel(channel))

        do while (handle(channel, reader.read() ?: break))
    }

    private fun getOptions(request: HTTPRequestMessage): HTTPResult<*> {
        val headers = ParameterList()
        headers["access-control-allow-origin"] = "*"
        headers["access-control-allow-methods"] =
            routes
                .filter { (_, value) -> value.any { it.route.matches(request.path) } }
                .map { it.key }
                .plusElement(HTTPMethod.OPTIONS)
                .joinToString(", ")
        request.headers["access-control-request-headers"]?.let {
            headers["access-control-allow-headers"] = it
        }
        headers["access-control-max-age"] = "3600"

        return HTTPResultUnit(
            204,
            "No Content",
            headers,
        )
    }

    private fun getArguments(
        request: HTTPRequestMessage,
        route: Route,
        parameters: List<KParameter>,
        arguments: Array<Any?>,
    ) {
        for (i in parameters.indices) {
            if (i == 0) {
                arguments[i] = route.instance
                continue
            }

            val parameter = parameters[i]

            val typename: String
            val value: Any?
            when {
                parameter.hasAnnotation<PathParameter>() -> {
                    val name = parameter.findAnnotation<PathParameter>()!!.value

                    typename = "path $name"
                    value = route.route.get(request.path, name)
                }

                parameter.hasAnnotation<QueryParameter>() -> {
                    val name = parameter.findAnnotation<QueryParameter>()!!.value
                    val values = request.query.getAll(name)

                    typename = "query $name"
                    value =
                        if (parameter.type.classifier == Array::class)
                            values.toTypedArray()
                        else
                            values.firstOrNull()
                }

                parameter.hasAnnotation<Header>() -> {
                    val name = parameter.findAnnotation<Header>()!!.value
                    val values = request.headers.getAll(name)

                    typename = "header $name"
                    value =
                        if (parameter.type.classifier == Array::class)
                            values.toTypedArray()
                        else
                            values.firstOrNull()
                }

                parameter.hasAnnotation<Body>() -> {
                    typename = "body"

                    value = request.body
                }

                else -> continue
            }

            if (value != null) {
                val type =
                    if (value::class == Array<String>::class)
                        typeOf<Array<String>>()
                    else
                        value::class.starProjectedType

                try {
                    arguments[i] = convert(
                        value,
                        type,
                        parameter.type,
                    )
                } catch (_: Exception) {
                    throw BadRequestSignal(content = "failed to convert parameter '$typename'")
                }
            } else if (!parameter.isOptional && !parameter.type.isMarkedNullable) {
                throw BadRequestSignal(content = "parameter '$typename' is neither optional nor nullable")
            }
        }
    }

    private fun getResult(request: HTTPRequestMessage): HTTPResult<*> {

        val candidates = routes
            .computeIfAbsent(request.method) { mutableListOf() }
            .stream()
            .filter { it.route.matches(request.path) }
            .max(Comparator.naturalOrder())

        if (candidates.isEmpty) {
            return NotFoundSignal().generate()
        }

        try {
            val route = candidates.get()

            val parameters = route.callee.parameters
            val arguments = arrayOfNulls<Any>(parameters.size)

            getArguments(
                request,
                route,
                parameters,
                arguments,
            )

            val value: Any?
            val type = route.callee.returnType

            try {
                value = route.callee.call(*arguments)
            } catch (e: InvocationTargetException) {
                throw e.targetException
            }

            if (value == null) {
                throw NullPointerException()
            }

            return convert(
                value,
                type,
                typeOf<HTTPResult<*>>(),
            ) as HTTPResult<*>
        } catch (s: Signal) {
            return s.generate()
        } catch (t: Throwable) {
            log.trace(t)
            return InternalServerErrorSignal().generate()
        }
    }

    private fun handle(channel: SocketChannel, request: HTTPRequestMessage): Boolean {
        log.info("${request.method} ${request.path} ${request.protocol}")

        val keepAlive = request.headers["connection"]?.lowercase() != "close"

        val result = when (request.method) {
            HTTPMethod.OPTIONS -> getOptions(request)
            else -> getResult(request)
        }

        val headers = ParameterList(result.headers)

        var chunked = false
        if (result !is HTTPResultUnit) {
            if ("content-type" !in headers) {
                headers["content-type"] = result.contentType
            }

            if ("content-length" !in headers && "transfer-encoding" !in headers) {
                chunked = result.count < 0

                if (chunked) {
                    headers["transfer-encoding"] = "chunked"
                } else {
                    headers["content-length"] = result.count.toString()
                }
            }
        }

        if ("access-control-allow-origin" !in headers) {
            headers["access-control-allow-origin"] = "*"
        }

        HTTPResponseMessage(
            "HTTP/1.1",
            result.statusCode,
            result.statusText,
            headers,
            if (result.channel != null)
                HTTPMessageBody(
                    result.channel,
                    result.position,
                    result.count,
                    chunked,
                )
            else null,
        ).use { it.write(channel) }

        return keepAlive
    }
}
